Objects/dictobject.c

```diff
@@ -140,6 +140,31 @@ To avoid slowing down lookups on a near-full table, we resize the table when
 it's USABLE_FRACTION (currently two-thirds) full.
 */
 
+#ifdef Py_GIL_DISABLED
+
+#define LOCK_KEYS(keys) PyMutex_Lock(&keys->dk_mutex)
```

Comment:
I think we want to use `_Py_LOCK_DONT_DETACH`. Otherwise, locking the keys object can release an outer critical section. This also means that it will be important that `LOCK_KEYS()` is essentially the "inner-most" code and that we don't call `Py_DECREF()` or similar while holding the keys lock. 


```suggestion
#define LOCK_KEYS(keys) PyMutex_LockFlags(&keys->dk_mutex, _Py_LOCK_DONT_DETACH)
```

Reply:
The only thing we ever decref is a split keys objec which will only hold unicode keys, but I'll move that out of the lock as it doesn't need to be under it anyway.